### PR TITLE
[FIX] stock: function field product_qty written in the new api.

### PR DESCRIPTION
**Pull request by simongoffin**
To avoid some problems with stored recomputed field in the old api.

### CODE REVIEW

File: addons/stock/stock.py [hunk 1]
```diff
@@ -1713,6 +1708,9 @@ def _set_product_qty(self, cr, uid, id, field, value, arg, context=None):
         """
         raise osv.except_osv(_('Programming Error!'), _('The requested operation cannot be processed because of a programming error setting the `product_qty` field instead of the `product_uom_qty`.'))
 
+    product_qty = new_fields.Float(compute='_quantity_normalize', inverse=_set_product_qty, digits=0, store=True, string='Quantity',
```
Comment (jco-odoo) [inline addons/stock/stock.py]: inverse is without ''?

**Comment by clonedagain** [comment]
Can you elaborate about the problems the current implementation has?

**Comment by jco-odoo** [comment]
Invoice before delivery

**Comment by odony** [comment]
@simongoffin Apparently this was committed at a1da6c2 before being reviewed, that doesn't sound right. In addition, changing a single field to the new API in the middle of an old-API class looks like it fixes the problem by chance rather than by design. Could you explain the rationale? Thanks!

**Comment by gurneyalex** [comment]
Thanks to @odony for pointing me to this PR.  

Before anything else: thanks for the quick fix (a1da6c2), the bug was a real show stopper. 

However I'm concerned that this is only a quick fix which does not solve the underlying cause. Here is my analysis:

I hit this issue when creating pickings from a SO with invoicing policy 'before delivery'. In which case, the call stack leading to the creation of the stock move by the procurement goes through several workflow transitions.

At one step of this stack, `account_move_line.reconcile()` is called, which triggers the creation of an `account_move_line`. This call to `models.create()` goes in the "recompute new-style fields" section of `models._create`, which calls `models.recompute`. The code of recompute was modified in c9154e0 by @beledouxdenis to use `with rec.env.norecompute()`.

As a result of this, when the `stock.move` is finally created, `recs.env.recompute` is `False`. This causes stored function fields to not be recomputed and therefore not stored (see https://github.com/odoo/odoo/blob/8.0/openerp/models.py#L4252) which is exactly the point of c9154e0 from the changeset description. 

It is not clear to me why new API function fields are not affected by this behavior (and therefore why the quick fix works). 

In my opinion, c9154e0 is a problematic changeset, as it introduces this regression, an potentially many others: from what I understand, any old api object created after the account move line in the chain of workflow propagation will suffer the same issues with its stored computed fields.

a1da6c2 does fix the problem for `stock.move`, but there are potentially many custom models affected by this, and possible other standard addons.

**Comment by gurneyalex** [comment]
OK, I just hit another manifestation of this. With the `delivery` addon installed, the fields `weight` and `weight_net` of `stock.move` are not computed when the delivery of a sale order is generated by the payment of the invoice. 

Steps to reproduce:

* set a non null weight, weight net and volume on product Ipod
* create a sale order for 10 ipods with create invoice before delivery
* confirm the SO, create the invoice, register payment of the invoice
* check the delivery for the SO

Expected: 
the stock.move have a non null weight and weight net

Actual: both are 0

**Comment by gurneyalex** [comment]
OPW 632624 opened.

**Pull request closed without merging**